kernel/mmanutils.c

```diff
@@ -92,35 +109,145 @@ int myst_teardown_mman(void)
+    uint64_t mman_start = (uint64_t)_mman_start;
+    size_t mman_size = _mman_size;
+    const uint64_t addr_end;
+    const size_t mman_end;
 
-    if (!(m = calloc(1, sizeof(msync_mapping_t))))
-        return NULL;
+    if (!addr || !length)
+        ERAISE(-EINVAL);
+
+    if (((uint64_t)addr % PAGE_SIZE) || (length % PAGE_SIZE))
+        ERAISE(-EINVAL);
+
+    if (__builtin_add_overflow(addr_start, length, &addr_end))
+        ERAISE(-ERANGE);
+
+    if (__builtin_add_overflow(mman_start, mman_size, &mman_end))
+        ERAISE(-ERANGE);
+
+    if (!(addr_start >= mman_start && addr_end <= mman_end))
+        ERAISE(-EINVAL);
 
-    m->fd = fd;
-    m->offset = offset;
-    m->addr = addr;
-    m->length = length;
+    ret = (addr_start - mman_start) / PAGE_SIZE;
 
-    return m;
+done:
+    return ret;
+}
+
+static void _free_fdmappings_pathnames(void* arg)
+{
+    uint8_t* addr = (uint8_t*)_mman.map;
+    size_t length = ((uint8_t*)_mman.end) - addr;
+    size_t index;
+    vectors_t v = _get_vectors();
+
+    (void)arg;
+
+    myst_round_up(length, PAGE_SIZE, &length);
+    index = _get_page_index(addr, length);
+    assert(index >= 0);
+    size_t count = length / PAGE_SIZE;
+
+    for (size_t i = index; i < index + count; i++)
+    {
+        myst_fdmapping_t* p = &v.fdmappings[i];
+
+        if (p->pathname)
+        {
+            myst_refstr_unref(p->pathname);
+            p->pathname = NULL;
+        }
+    }
+}
+
+static myst_once_t _free_fdmappings_pathnames_atexit_once;
+
+static void _free_fdmappings_pathnames_atexit(void)
+{
+    myst_atexit(_free_fdmappings_pathnames, NULL);
+}
+
+static int _add_file_mapping(int fd, off_t offset, void* addr, size_t length)
+{
+    int ret = 0;
+    size_t index;
+    vectors_t v = _get_vectors();
+    struct locals
+    {
+        char pathname[PATH_MAX];
+    };
+    struct locals* locals = NULL;
+    myst_refstr_t* pathname = NULL;
+
+    if (fd < 0 || offset < 0 || !addr || !length)
+        ERAISE(-EINVAL);
+
+    if (!(locals = calloc(1, sizeof(struct locals))))
+        ERAISE(-ENOMEM);
+
+    /* register the cleanup function for fd-mapping pathnames with atxit() */
+    myst_once(
+        &_free_fdmappings_pathnames_atexit_once,
+        _free_fdmappings_pathnames_atexit);
+
+    ECHECK(_fd_to_pathname(fd, locals->pathname));
+
+    /* make a reference-counted version of the pathname */
+    if (!(pathname = myst_refstr_dup(locals->pathname)))
+        ERAISE(-ENOMEM);
+
+    ECHECK(myst_round_up(length, PAGE_SIZE, &length));
+    ECHECK((index = _get_page_index(addr, length)));
+
+    myst_rspin_lock(&_mman.lock);
+    {
+        const size_t count = length / PAGE_SIZE;
+        uint64_t off = offset;
+
+        for (size_t i = index; i < index + count; i++)
+        {
+            myst_fdmapping_t* p = &v.fdmappings[i];
+
+            /* ATTN: the entry might already be in use for unknown reasons */
+            if (p->used == MYST_FDMAPPING_USED)
```

Comment:
how? Someone is trying to map to an address someone else is using? Is this code hit?

Reply:
Unresolved! I need to think more about how this could happen.

Reply:
I added a comment to the code to explain.